File: algorithms/0003.longest-substring-without-repeating-characters/solution.rs
use std::cmp::max;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut curr: Vec<char> = Vec::new();
        let mut longest = 0;

        for c in s.chars() {
            if let Some(i) = curr.iter().position(|&x| x == c) {
                curr.drain(0..(i + 1));
            }

            curr.push(c);
            longest = max(longest, curr.len());
        }

        longest as i32
    }
}

File: algorithms/0003.longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
  let curr = [];
  let longest = 0;

  for (const c of s) {
    const idx = curr.indexOf(c);
    if (idx >= 0) {
      curr = curr.slice(idx + 1);
    }

    curr.push(c);
    if (curr.length > longest) {
      longest = curr.length;
    }
  }

  return longest;
}
